File: lingting-grpc-client/src/main/java/live/lingting/framework/grpc/client/GrpcChannelBuilder.java
package live.lingting.framework.grpc.client;

import io.grpc.ClientInterceptor;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.AbstractAsyncStub;
import io.grpc.stub.AbstractBlockingStub;
import io.grpc.stub.AbstractFutureStub;
import io.grpc.stub.AbstractStub;
import live.lingting.framework.grpc.client.properties.GrpcClientProperties;
import live.lingting.framework.util.StringUtils;
import lombok.RequiredArgsConstructor;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Function;
import java.util.function.UnaryOperator;

/**
 * @author lingting 2024-01-30 13:55
 */
@RequiredArgsConstructor
public class GrpcChannelBuilder {

	protected final GrpcClientProvide provide;

	private final List<ClientInterceptor> interceptors = new ArrayList<>();

	protected String target;

	protected GrpcClientProperties properties;

	public GrpcChannelBuilder address(String host, Integer port) {
		if (!StringUtils.hasText(host)) {
			throw new IllegalArgumentException("Host [%s] is invalid!".formatted(host));
		}
		if (port == null || port < 0 || port > 65535) {
			throw new IllegalArgumentException("Port [%d] is invalid!".formatted(port));
		}
		this.target = "%s:%d".formatted(host, port);
		return this;
	}

	public GrpcChannelBuilder target(String target) {
		this.target = target;
		return this;
	}

	public GrpcChannelBuilder interceptor() {
		return interceptor(provide.interceptors);
	}

	public GrpcChannelBuilder interceptor(ClientInterceptor interceptor) {
		interceptors.add(interceptor);
		return this;
	}

	public GrpcChannelBuilder interceptor(Collection<ClientInterceptor> collection) {
		interceptors.addAll(collection);
		return this;
	}

	/**
	 * 使用 provide的配置填充
	 */
	public GrpcChannelBuilder provide() {
		return interceptor().properties();
	}

	public GrpcChannelBuilder properties() {
		return properties(provide.properties);
	}

	public GrpcChannelBuilder properties(GrpcClientProperties properties) {
		this.properties = properties;
		return this;
	}

	public ManagedChannel build() {
		return build((UnaryOperator<ManagedChannelBuilder<?>>) builder -> builder);
	}

	public ManagedChannel build(UnaryOperator<ManagedChannelBuilder<?>> operator) {
		return build((Function<String, ManagedChannelBuilder<?>>) forTarget -> {
			ManagedChannelBuilder<?> builder = ManagedChannelBuilder.forTarget(forTarget);
			return operator.apply(builder);
		});
	}

	public ManagedChannel build(Function<String, ManagedChannelBuilder<?>> function) {
		ManagedChannelBuilder<?> builder = function.apply(target);
		if (properties != null) {
			provide.useProperties(builder, properties);
		}
		provide.addInterceptors(builder, interceptors);
		return builder.build();
	}

	public <R extends AbstractStub<R>> R stub(Function<ManagedChannel, R> function) {
		ManagedChannel channel = build();
		return function.apply(channel);
	}

	public <R extends AbstractAsyncStub<R>> R async(Function<ManagedChannel, R> function) {
		return stub(function);
	}

	public <R extends AbstractBlockingStub<R>> R blocking(Function<ManagedChannel, R> function) {
		return stub(function);
	}

	public <R extends AbstractFutureStub<R>> R future(Function<ManagedChannel, R> function) {
		return stub(function);
	}

}

File: lingting-grpc/src/main/kotlin/live/lingting/framework/grpc/GrpcChannelBuilder.kt
package live.lingting.framework.grpc

import io.grpc.ClientInterceptor
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.stub.AbstractAsyncStub
import io.grpc.stub.AbstractBlockingStub
import io.grpc.stub.AbstractFutureStub
import io.grpc.stub.AbstractStub
import java.util.function.Function
import java.util.function.UnaryOperator
import live.lingting.framework.grpc.properties.GrpcClientProperties

/**
 * @author lingting 2024-01-30 13:55
 */
class GrpcChannelBuilder(val provide: GrpcClientProvide) {

    private val interceptors: MutableList<ClientInterceptor> = ArrayList()

    var target: String? = null

    var properties: GrpcClientProperties? = null

    fun address(host: String, port: Int): GrpcChannelBuilder {
        require(host.isNotEmpty()) { "Host [$host] is invalid!" }
        require(!(port < 0 || port > 65535)) { "Port [$port] is invalid!" }
        return target("$host:$port")
    }

    fun target(target: String): GrpcChannelBuilder {
        this.target = target
        return this
    }

    fun interceptor(): GrpcChannelBuilder {
        return interceptor(provide.interceptors)
    }

    fun interceptor(interceptor: ClientInterceptor): GrpcChannelBuilder {
        interceptors.add(interceptor)
        return this
    }

    fun interceptor(collection: Collection<ClientInterceptor>): GrpcChannelBuilder {
        interceptors.addAll(collection)
        return this
    }

    /**
     * 使用 provide的配置填充
     */
    fun provide(): GrpcChannelBuilder {
        return interceptor().properties()
    }

    fun properties(): GrpcChannelBuilder {
        return properties(provide.properties)
    }

    fun properties(properties: GrpcClientProperties): GrpcChannelBuilder {
        this.properties = properties
        return this
    }

    @JvmOverloads
    fun build(operator: UnaryOperator<ManagedChannelBuilder<*>> = UnaryOperator { builder: ManagedChannelBuilder<*> -> builder }): ManagedChannel {
        return build(Function { forTarget ->
            val builder = ManagedChannelBuilder.forTarget(forTarget)
            operator.apply(builder)
        })
    }

    fun build(function: Function<String, ManagedChannelBuilder<*>>): ManagedChannel {
        val builder = target.let {
            require(!it.isNullOrBlank()) { "Target is invalid!" }
            function.apply(it)
        }
        properties?.let { provide.useProperties(builder, it) }
        provide.addInterceptors(builder, interceptors)
        return builder.build()
    }

    fun <R : AbstractStub<R>> stub(function: Function<ManagedChannel, R>): R {
        val channel = build()
        return function.apply(channel)
    }

    fun <R : AbstractAsyncStub<R>> async(function: Function<ManagedChannel, R>): R {
        return stub<R>(function)
    }

    fun <R : AbstractBlockingStub<R>> blocking(function: Function<ManagedChannel, R>): R {
        return stub<R>(function)
    }

    fun <R : AbstractFutureStub<R>> future(function: Function<ManagedChannel, R>): R {
        return stub<R>(function)
    }
}
